src/transaction.rs

```diff
@@ -1,78 +1,360 @@
+use crate::circuit::blinding_circuit::BlindingCircuit;
+use crate::circuit::circuit_parameters::CircuitParameters;
+use crate::circuit::validity_predicate::{ValidityPredicate, NUM_NOTE};
+use crate::constant::{
+    ACTION_CIRCUIT_SIZE, ACTION_PUBLIC_INPUT_CM_INDEX, ACTION_PUBLIC_INPUT_NF_INDEX,
+    ACTION_PUBLIC_INPUT_ROOT_INDEX, BLINDING_CIRCUIT_SIZE,
 };
-use ark_ec::twisted_edwards_extended::GroupAffine as TEGroupAffine;
-use rs_merkle::{algorithms::Blake2s, MerkleTree};
-
-pub struct Transaction<'a, CP: CircuitParameters> {
-    //max: usize, // the maximum number of actions/notes for a transaction
-    actions: Vec<Action<CP>>,
-    spent_notes: Vec<(Note<CP>, Nullifier<CP>)>,
-    created_notes: Vec<(Note<CP>, EncryptedNote<CP::InnerCurve>)>,
-    vps: &'a Vec<ValidityPredicate<CP>>,
+use crate::error::TaigaError;
+use crate::vp_description::ValidityPredicateDescription;
+use plonk_core::circuit::Circuit;
+use plonk_core::circuit::{verify_proof, VerifierData};
+use plonk_core::proof_system::{pi::PublicInputs, Proof, VerifierKey};
+use rand::RngCore;
+
+pub const NUM_TX_SLICE: usize = NUM_NOTE;
+
+pub struct Transaction<CP: CircuitParameters> {
+    pub action_slices: [ActionSlice<CP>; NUM_TX_SLICE],
+    pub spend_slices: [SpendSlice<CP>; NUM_TX_SLICE],
+    pub output_slices: [OutputSlice<CP>; NUM_TX_SLICE],
 }
 
-impl<'a, CP: CircuitParameters> Transaction<'a, CP> {
-    pub fn new(
-        //max: usize,
-        actions: Vec<Action<CP>>,
-        spent_notes: Vec<(Note<CP>, Nullifier<CP>)>,
-        created_notes: Vec<(Note<CP>, EncryptedNote<CP::InnerCurve>)>,
-        vps: &'a Vec<ValidityPredicate<CP>>,
-    ) -> Self {
+pub struct ActionSlice<CP: CircuitParameters> {
+    pub action_proof: Proof<CP::CurveScalarField, CP::CurvePC>,
+    pub action_public: Action<CP>,
+}
+
+pub struct SpendSlice<CP: CircuitParameters> {
+    spend_addr_vp: VPCheck<CP>,
+    spend_token_vp: VPCheck<CP>,
+}
+
+pub struct OutputSlice<CP: CircuitParameters> {
+    output_addr_vp: VPCheck<CP>,
+    output_token_vp: VPCheck<CP>,
+}
+
+pub struct VPCheck<CP: CircuitParameters> {
+    // VP circuit size
+    pub vp_circuit_size: usize,
+    // VP proof
+    pub vp_proof: Proof<CP::CurveScalarField, CP::CurvePC>,
+    // The public inputs for vp proof
+    // TODO: maybe the PublicInputs<CP::CurveScalarField> should be serialized from Vec<CP::CurveScalarField>, like the action_public in ActionSlice?
+    pub vp_public_inputs: PublicInputs<CP::CurveScalarField>,
+    // The vk of vp is blinded partially, blinded parts can be constructed from blinded_vp_public_inputs.
+    // The unblinded_vp_vk is the unblinded parts.
+    // TODO:
+    pub unblinded_vp_vk: VerifierKey<CP::CurveScalarField, CP::CurvePC>,
+    // blind vp proof
+    pub blind_vp_proof: Proof<CP::CurveBaseField, CP::OuterCurvePC>,
+    // The public inputs for blind vp proof
+    pub blinded_vp_public_inputs: PublicInputs<CP::CurveBaseField>,
+    // TODO:
```

Comment:
I don't get what it is... but it is a TODO :-D

Reply:
1. Some public inputs are needed to check between the proofs. I don't want to handle the magic variable indexes now, since our implementation are changing. Let's fix this later.
2. After handling the magic variable indexes, we can replace `PublicInputs<F>` by `Vec<F>`. I think It's more reasonable.
3. Some parameters in VP are need to be designed.